src/lucky/memoizable.cr

```diff
@@ -15,25 +15,69 @@ module Lucky::Memoizable
   # then each subsequent call returns the user record.
   #
   # The `memoize` method will raise a compile time exception if you forget to include
-  # a return type for your method, or if your return type is a `Union`.
-  # Arguments are not allowed in memoized methods because these can change the return value.
+  # a return type for your method, or if any arguments are missing a type.
+  # Arguments are allowed but as soon as one changes, the previous value is no longer held on to.
+  # Equality (==) is used for checking on argument updates.
   macro memoize(method_def)
-    {% raise "Return type of memoize method must not be a Union" if method_def.return_type.is_a?(Union) %}
-    {% raise "You must define a return type for memoize methods" if method_def.return_type.is_a?(Nop) %}
-    {% raise "Memoize methods can not be defined with arguments" if method_def.args.size > 0 %}
+    {% raise "You must define a return type for memoized methods" if method_def.return_type.is_a?(Nop) %}
+    {% raise "All arguments must have an explicit type for memoized methods" if method_def.args.any? &.is_a?(Nop) %}
 
-    @__{{ method_def.name }} : {{ method_def.return_type }}? = nil
+    @__memoized_{{method_def.name}} : Tuple(
+      {{ method_def.return_type }},
+      {% for arg in method_def.args %}
+        {{ arg.restriction }},
+      {% end %}
+    )?
 
     # Returns uncached value
-    def {{ method_def.name }}__uncached : {{ method_def.return_type }}
+    def {{ method_def.name }}__uncached(
+      {% for arg in method_def.args %}
+        {{ arg.name }} : {{ arg.restriction }},
+      {% end %}
+    ) : {{ method_def.return_type }}
       {{ method_def.body }}
     end
 
-    # Returns cached value
-    def {{ method_def.name }} : {{ method_def.return_type }}
-      @__{{ method_def.name }} ||= -> do
-        {{ method_def.name }}__uncached
+    def {{ method_def.name }}__tuple_cached(
```

Comment:
Can we add a comment above this like the others? Maybe "Stores the cache in the tuple" or something?